sage.py

```diff
@@ -481,1274 +423,920 @@ def getepisodes(action_seq, mcat, plot, debug=False):
+print('------ !! Special: Fixing syntax error in main model and sink files ------')
 print('--- Sinks')
-with open(path_to_model+".ff.finalsinks.json", 'r') as file:
+with open(path_to_traces + ".ff.finalsinks.json", 'r') as file:
     filedata = file.read()
-stripped = re.sub('[\s+]', '', filedata)
-extracommas = re.search('(}(,+)\]}$)', stripped)
-if extracommas is not None:
-    c = (extracommas.group(0)).count(',')
-    print(extracommas.group(0), c)
-    filedata = ''.join(filedata.rsplit(',', c))
-    with open(path_to_model+".ff.finalsinks.json", 'w') as file:
+stripped = re.sub(r'[\s+]', '', filedata)
+extra_commas = re.search(r'(}(,+)]}$)', stripped)
+if extra_commas is not None:
+    comma_count = (extra_commas.group(0)).count(',')
+    print(extra_commas.group(0), comma_count)
+    filedata = ''.join(filedata.rsplit(',', comma_count))
+    with open(path_to_traces + ".ff.finalsinks.json", 'w') as file:
         file.write(filedata)
-  
-print('--- Main')  
-with open(path_to_model+".ff.final.json", 'r') as file:
+
+print('--- Main')
+with open(path_to_traces + ".ff.final.json", 'r') as file:
     filedata = file.read()
-stripped = re.sub('[\s+]', '', filedata)
-extracommas = re.search('(}(,+)\]}$)', stripped)
-if extracommas is not None:
-    c = (extracommas.group(0)).count(',')
-    print(extracommas.group(0), c)
-    filedata = ''.join(filedata.rsplit(',', c))
-    with open(path_to_model+".ff.final.json", 'w') as file:
+stripped = re.sub(r'[\s+]', '', filedata)
+extra_commas = re.search(r'(}(,+)]}$)', stripped)
+if extra_commas is not None:
+    comma_count = (extra_commas.group(0)).count(',')
+    print(extra_commas.group(0), comma_count)
+    filedata = ''.join(filedata.rsplit(',', comma_count))
+    with open(path_to_traces + ".ff.final.json", 'w') as file:
         file.write(filedata)
 
-print('------ Loading and traversing SPDFA ---------')
-# Load S-PDFA
-m, data = loadmodel(path_to_model+".ff.final.json")
-m2,data2 = loadmodel(path_to_model+".ff.finalsinks.json")
+print('------ Loading and traversing S-PDFA ------')
+main_model = load_model(path_to_traces + ".ff.final.json")
+sinks_model = load_model(path_to_traces + ".ff.finalsinks.json")
+
+print('------ Encoding traces into state sequences ------')
+state_traces, med_sev_states, high_sev_states, severe_sinks = encode_sequences(main_model, sinks_model, episode_traces)
+state_sequences = make_state_sequences(episode_subsequences, state_traces, med_sev_states, high_sev_states)
 
-print('------- Encoding into state sequences --------')
-# Encoding traces into state sequences  
-(traces, state_traces) = encode_sequences(m,m2)
-(med_states, sev_states, sev_sinks) = find_severe_states(traces, m, m2)    
-condensed_data = make_condensed_data(alerts, keys, state_traces, med_states, sev_states)
+# print('------ Clustering state groups ------')
+# state_groups = make_state_groups(state_sequences, path_to_traces)
```

Comment:
I have commented this out, as it is not really needed for AGs and can always be uncommented if needed